include/exadg/operators/inverse_mass_operator.h

```diff
@@ -81,16 +83,20 @@ class InverseMassOperator
 
     dealii::FiniteElement<dim> const & fe = matrix_free->get_dof_handler(dof_index).get_fe();
 
-    if(fe.conforms(dealii::FiniteElementData<dim>::L2))
+    // The inverse mass operator is only available for discontinuous Galerkin discretizations
+    AssertThrow(fe.conforms(dealii::FiniteElementData<dim>::L2),
+                dealii::ExcMessage("InverseMassOperator only implemented for DG!"));
+
+    // Currently, the inverse mass realized as matrix-free operator evaluation is only available
+    // in deal.II for tensor-product elements
+    if(fe.base_element(0).dofs_per_cell == dealii::Utilities::pow(fe.degree + 1, dim))
     {
-      // this checks if we have a tensor-product element
-      if(fe.base_element(0).dofs_per_cell == dealii::Utilities::pow(fe.degree + 1, dim))
-        explicit_matrix_free_inverse_mass_available = true;
+      inverse_mass_available_as_matrix_free_operator = true;
```

Comment:
Let me add a (currently only theoretical) note of what we had discussed at some point in the past: Even if we have a tensor product of homogeneous degree, we might still want to do an iterative solution, e.g. because we have a (strongly deformed) geometry such that the evaluation of the mass operator is done with more quadrature points than `p+1` (or `k+1`). Since the inverse is strongly linked to `n_q_points_1d == p + 1`, we would then need to have different quadrature formulas for the forward evaluation and the inverse, and might want to use the iterative solver to gain this accuracy of the more accurate integration.

Reply:
This comment does not mean that there is something specific to be changed, but it might be useful to reflect this in a comment somewhere that it is not a strict separation between "inverse via tensor product available" and "iterative inversion".

Reply:
Do you mean we should extend the above if-statement and check if the number of quadrature points is `n_q_points_1d == p + 1`? Otherwise, we would set `inverse_mass_available_as_matrix_free_operator = false` and use the iterative solution similar to the simplex case.

Another option would be to hand the responsibility over to the user and let the user specify which method to choose. This would allow to e.g. use the iterative solver also in cases where the explicit tensor-product inverse would principally be available (currently, to run this version (e.g. for performance comparisons), one would need to "hack" the code). When deciding for this option, the implementation of the inverse mass operator would have to assert that the user does not set `inverse_mass_available_as_matrix_free_operator = true` is cases where deal.II does actually not support this. Deciding for this option would imply having *more control* and using *less automation*, but I would also agree to such an option.

Reply:
I would personally prefer to have control, especially as you mention the performance tests. On the other hand, I'm wondering if the number of parameters to set would become overwhelming at some point - we talk about a single one here, to be set in case the default to use the matrix-free operator is not available. I think we can then proceed as you suggest, and assert in case the matrix-free operator is selected.

Reply:
Thinking more about this, I think one would need an enum

```c++
enum class
InverseMassType
{
  MatrixfreeOperator,
  ElementwiseKrylovSolver,
  BlockMatrices
};
````
instead of the boolean parameter
```c++
bool implement_block_diagonal_solver_matrix_free;
````

The second and third option of the above enum are available for all DG elements. In case the first option is set by the user, we need to assert that the current restrictions of the cellwise inverse mass operator in deal.II are met, in order to have useful error messages for ExaDG users. As a default value, I would initialize this variable with `InverseMassType::MatrixfreeOperator`. It is not the most general version, but the one that is used most frequently. I think this is better than using one of the other options as default, with the consequence that many users will not be aware of the fact that a faster option would be available for `ElementType::Hypercube`.

@kronbichler Do you agree?

Reply:
Yes, very good choice for both the enum and the default choice!